projects/hathor-wallet-headless/0004-tx-transaction.md

```diff
@@ -0,0 +1,663 @@
+The mint deposit inputs are manually added to the transaction and all outputs are shuffled except for the data output at position 0.
+
+```json
+[
+  { "type": "input/utxo", "position": -1, "fill": 2, "token": "00", "autoChange": false },
+  { "type": "input/utxo", "position": -1, "fill": 1, "token": "TST", "authority": "mint", "autoChange": false },
+  { "type": "output/token", "position": -1, "token": "TST", "amount": 1 },
+  { "type": "output/token", "position": -1, "token": "TST", "authority": "mint", "amount": 1 },
+  { "type": "action/change" },
+  { "type": "action/shuffle", "target": "outputs" },
+  { "type": "output/data", "position": 0, "data": "foobar" },
+]
+```
+
+This same transaction template instructions set could be built as:
+
+```ts
+// Token UID for token TST
+let TST = '000cafe...cafe123';
+
+const ttInstructions = new TransactionTemplateBuilder()
+  .addUtxoInput(-1, 2)
+  .addUtxoInput(-1, 2, TST, "mint")
+  .addTokenOutput(-1, 1, TST)
+  .addTokenOutput(-1, 1, TST, "mint")
+  .addShuffleAction("outputs")
+  .addDataOutput("foobar")
+  .export();
+```
+
+We can get a transaction template instance from the instructions set by the `build` or by calling the transaction template `from` with the transaction template instructions:
+
+```ts
+const tt = TransactionTemplate.from(ttInstructions);
+```
+
+Then we can get the designed transaction by calling the `build`:
+
+```ts
+const transaction = tt.build(myWalletFacade);
+```
+
+## Interpreters
+
+A class which responsibility is to read the sequence of instructions to create a transaction using a determined type of wallet.
+
+```ts
+interface TxTemplateInstructionInterpreter {
+  execute(tt: TransactionTemplate): Transaction;
+}
+```
+
+- execute: Will create a `TxTemplateInterpreterContext` instance and execute each instruction with the context then return the resulting `Transaction` instance.
+
+We have two instances of interpreters:
+
+- WalletFacadeTemplateInterpreter
+- ServiceTemplateInterpreter
+
+```ts
+interface TxTemplateInterpreterContext {
+  version: number,
+  signalBits: number,
+  inputs: Input[],
+  outputs: Output[],
+  tokens: string[],
+  balance: Record<string, IBalance>,
+  wallet: IHathorWallet,
+  isWalletService: boolean,
+  vars: Record<string, string|number>,
+}
+```
+
+## Lib land
+
+Beyond the wallet API we can use any utils or helper available to support the interpreter in its mission.
+
+For now we don't need to create anything. However, it worths to mention the `IHathorWallet` doesn't conveys all methods available on both wallets. Just to reference one, we can take the `getUtxos` which is implemented in both wallets but is not present at `IHathorWallet`. Not to mention the lack of type inference at wallet facade which makes the development experience poorer.
+
+>[!NOTE]
+>Decision Call: Should we use this implementation as an opportunity to improve the lib in the topics mentioned?
```

Comment:
suggestion: Add a number to the decision to keep with the pattern so far on this doc.
```suggestion
>Decision 5: Should we use this implementation as an opportunity to improve the lib in the topics mentioned?
```